### PR TITLE
Add SACA framework Op and SACA use case

### PR DESCRIPTION
**Pull request by elevans**
This PR adds the SACA framework (3 Ops: z-score heat map, p-value heatmap, sig mask) and the use case that goes with it.

### CODE REVIEW

File: scijava-ops-image/templates/main/java/org/scijava/ops/image/simplify/RAIFocusers.vm [hunk 1]
```diff
@@ -0,0 +1,101 @@
+/*
+ * #%L
```
Comment (gselzer) [inline scijava-ops-image/templates/main/java/org/scijava/ops/image/simplify/RAIFocusers.vm]: Can we delete this?

File: scijava-ops-image/templates/test/java/org/scijava/ops/image/simplify/RAIFocusersTest.vm [hunk 2]
```diff
@@ -0,0 +1,72 @@
+/*
+ * #%L
+ * ImageJ2 software for multidimensional image processing and analysis.
```
Comment (gselzer) [inline scijava-ops-image/templates/test/java/org/scijava/ops/image/simplify/RAIFocusersTest.vm]: Can we delete this?

File: scijava-ops-image/templates/main/java/org/scijava/ops/image/simplify/RAIFocusers.list [hunk 3]
```diff
@@ -0,0 +1,25 @@
+.include templates/main/java/org/scijava/ops/image/adapt/Globals.list
+
+[RAIFocusers.java]
```
Comment (gselzer) [inline scijava-ops-image/templates/main/java/org/scijava/ops/image/simplify/RAIFocusers.list]: Can we delete this?

File: scijava-ops-image/templates/test/java/org/scijava/ops/image/simplify/RAIFocusersTest.list [hunk 4]
```diff
@@ -0,0 +1,23 @@
+.include templates/main/java/org/scijava/ops/image/adapt/Globals.list
+
+[RAIFocusersTest.java]
+
+types = ```
```
Comment (gselzer) [inline scijava-ops-image/templates/test/java/org/scijava/ops/image/simplify/RAIFocusersTest.list]: Can we delete this?

File: scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java [hunk 5]
```diff
@@ -670,10 +670,17 @@ private List<RichOp<?>> resolveOpDependencies(OpInfo info,
 	private OpRequest inferOpRequest(OpDependencyMember<?> dependency,
 		Map<TypeVariable<?>, Type> typeVarAssigns)
 	{
```
Comment (gselzer) [inline scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java]: Is this change necessary?

File: scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java [hunk 6]
```diff
@@ -670,10 +670,17 @@ private List<RichOp<?>> resolveOpDependencies(OpInfo info,
...
+		final Map<TypeVariable<?>, Type> inferences = new HashMap<>(typeVarAssigns);
+		for (var entry : inferences.entrySet()) {
+			var value = entry.getValue();
+			if (value.equals(Any.class) || value instanceof Any) {
```
Comment (gselzer) [inline scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java]: If these changes *are* necessary, I wrote `Any.is()` to simplify this check

File: scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/adapt/AdaptationMatchingRoutine.java [hunk 7]
```diff
@@ -293,7 +293,7 @@ private OpRequest inferOpRequest(Type type, String name,
 
 	@Override
 	public double priority() {
-		return Priority.LOW;
+		return Priority.VERY_LOW;
```
Comment (gselzer) [inline scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/adapt/AdaptationMatchingRoutine.java]: Is this change necessary?

File: scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/AdaptiveSmoothedKendallTau.java [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+ *
+ * @author Shulei Wang
+ * @author Ellen TA Dobson
+ */
+
+public final class AdaptiveSmoothedKendallTau {
+
+	private AdaptiveSmoothedKendallTau() {}
+
+	// TODO: check that output float type is actually what we want here.
```
Comment (gselzer) [inline scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/AdaptiveSmoothedKendallTau.java]: Is this `TODO` still valid?

File: scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/PNorm.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+ * of pixelwise p-values. This class replicates R's pnorm function.
+ *
+ * @author Edward Evans
+ */
+
+public final class PNorm {
+
+	private PNorm() {}
+
+	public static void compute(RandomAccessibleInterval<DoubleType> input,
```
Comment (gselzer) [inline scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/PNorm.java]: I feel like this could be its own Op

File: scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/QNorm.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+ * Helper class for Spatially Adaptive Colocalization Analysis (SACA) framework.
+ * This class is used by the "coloc.saca.sigMask" Op to produce a binary mask of
+ * significantly colocalized pixels. This class replicates R's qnorm function.
+ *
+ * @author Shulei Wang
+ * @author Ellen TA Dobson
+ * @author Curtis Rueden
+ */
+
+public final class QNorm {
```
Comment (gselzer) [inline scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/QNorm.java]: These could also be simple Op methods

File: scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/SACAHeatmapPValue.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * @author Shulei Wang
+ * @author Curtis Rueden
+ * @author Ellen TA Dobson
+ * @author Edward Evans
+ * @implNote op names='coloc.saca.heatmapPValue', priority='100.'
+ */
+
+public class SACAHeatmapPValue implements
```
Comment (gselzer) [inline scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/SACAHeatmapPValue.java]: This is fine as is, but if you didn't want to implement the `Computers.Arity2` interface you could easily make it a static method.

File: scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/WtKendallTau.java [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+/**
+ * Helper class for Spatially Adaptive Colocalization Analysis (SACA) framework.
+ * This class is used by the AdaptiveSmoothedKendallTau class.
+ *
+ * @author Shulei Wang
+ * @author Ellen TA Dobson
+ * @author Curtis Rueden
+ */
+
+final class WtKendallTau {
```
Comment (gselzer) [inline scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/WtKendallTau.java:48]: Could also consider making these their own ops

Comment (gselzer) [reply scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/WtKendallTau.java:48]: This one I feel less strongly about, because I do not know whether these would be used elsewhere

File: scijava-ops-image/src/main/java/org/scijava/ops/image/simplify/RAIFocusers.java [hunk 13]
```diff
@@ -0,0 +1,227 @@
+///*
+// * #%L
+// * ImageJ2 software for multidimensional image processing and analysis.
+// * %%
+// * Copyright (C) 2014 - 2023 ImageJ2 developers.
+// * %%
+// * Redistribution and use in source and binary forms, with or without
+// * modification, are permitted provided that the following conditions are met:
+// *
+// * 1. Redistributions of source code must retain the above copyright notice,
```
Comment (gselzer) [inline scijava-ops-image/src/main/java/org/scijava/ops/image/simplify/RAIFocusers.java]: I'd delete this

File: scijava-ops-image/src/test/java/org/scijava/ops/image/coloc/saca/SACATest.java [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+import net.imglib2.type.numeric.real.DoubleType;
+import net.imglib2.type.numeric.integer.UnsignedByteType;
+
+import org.scijava.types.Nil;
+import org.scijava.ops.image.AbstractColocalisationTest;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/**
+ * Tests {@code coloc.saca} ops.
```
Comment (gselzer) [inline scijava-ops-image/src/test/java/org/scijava/ops/image/coloc/saca/SACATest.java:46]: Add yourself as an author

File: scijava-ops-image/src/test/java/org/scijava/ops/image/coloc/saca/SACATest.java [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+ */
+
+public class SACATest extends AbstractColocalisationTest {
+
+	@Test
+	public void testSACAFramework() {
```
Comment (gselzer) [inline scijava-ops-image/src/test/java/org/scijava/ops/image/coloc/saca/SACATest.java]: I'd split these into 3 separate tests for granularity and readability. i.e. it's nice to see quickly which of the Ops fail if any fail.

**Review by gselzer — CHANGES_REQUESTED** [review]
@elevans by and large this looks great! Thank you for porting this - just a couple small changes, none of which I'd expect to take any real work.

The one strange thing here is that there are edits to `RAIFocusers.java` and related templates which are probably unnecessary. Can we see if this PR still works without those changes? Same goes for the edits to SciJava Ops Engine changes - I think I put comments on each

**Comment by elevans** [comment]
@gselzer Absolutely! So the `RAIFocusers` commit came from a rebase that had merge conflicts. I probably messed this up. So if this works without these changes, Great!!

File: scijava-ops-image/templates/main/java/org/scijava/ops/image/simplify/RAIFocusers.vm (hunk 1 above)
Comment (elevans) [reply scijava-ops-image/templates/main/java/org/scijava/ops/image/simplify/RAIFocusers.vm]: Done! 8e987ba24e31c3d998bd9e7e102a1e20f9779528.

File: scijava-ops-image/templates/test/java/org/scijava/ops/image/simplify/RAIFocusersTest.vm (hunk 2 above)
Comment (elevans) [reply scijava-ops-image/templates/test/java/org/scijava/ops/image/simplify/RAIFocusersTest.vm]: Done! 8e987ba24e31c3d998bd9e7e102a1e20f9779528.

File: scijava-ops-image/templates/main/java/org/scijava/ops/image/simplify/RAIFocusers.list (hunk 3 above)
Comment (elevans) [reply scijava-ops-image/templates/main/java/org/scijava/ops/image/simplify/RAIFocusers.list]: Done! 8e987ba24e31c3d998bd9e7e102a1e20f9779528.

File: scijava-ops-image/templates/test/java/org/scijava/ops/image/simplify/RAIFocusersTest.list (hunk 4 above)
Comment (elevans) [reply scijava-ops-image/templates/test/java/org/scijava/ops/image/simplify/RAIFocusersTest.list]: Done! 8e987ba24e31c3d998bd9e7e102a1e20f9779528.

File: scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java (hunk 5 above)
Comment (elevans) [reply scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java]: No it's not. I tried looking to see where these changes came from but its a bit murky. My best guess is that I rebased from an experimental `scijava-ops-engine` branch that you were still working on and I just carried it over. Anyways, we definitely don't need this. Fixed here: aae10fb4c6e1edd09da852d3b9208cd2bc09323a.

File: scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java (hunk 6 above)
Comment (elevans) [reply scijava-ops-engine/src/main/java/org/scijava/ops/engine/impl/DefaultOpEnvironment.java]: The changes are necessary. Fixed here: aae10fb4c6e1edd09da852d3b9208cd2bc09323a

File: scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/adapt/AdaptationMatchingRoutine.java (hunk 7 above)
Comment (elevans) [reply scijava-ops-engine/src/main/java/org/scijava/ops/engine/matcher/adapt/AdaptationMatchingRoutine.java]: No it's not. `Priority.LOW` works just fine! Fixed: aae10fb4c6e1edd09da852d3b9208cd2bc09323a

File: scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/SACAHeatmapPValue.java (hunk 11 above)
Comment (hinerm) [reply scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/SACAHeatmapPValue.java]: I know we discussed today but wanted to note that this class will be deleted in favor of the opified PNorm

**Comment by hinerm** [comment]
I don't see anything that needs revision beyond what Gabe said. Thanks @elevans!

**Comment by elevans** [comment]
Thanks! Almost done!

File: scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/SACAHeatmapPValue.java (hunk 11 above)
Comment (elevans) [reply scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/SACAHeatmapPValue.java]: Correct! This class has been deleted.

File: scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/AdaptiveSmoothedKendallTau.java (hunk 8 above)
Comment (elevans) [reply scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/AdaptiveSmoothedKendallTau.java]: It is not, good catch. I checked the other `TODO`s there is only 1 left that might need a look at later down the road (swaping Tau values instead of copying them). Fixed in 72d2d7dc14cc0691b67d23aca3089df7aad10bf1

File: scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/PNorm.java (hunk 9 above)
Comment (elevans) [reply scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/PNorm.java]: Agreed! It's now its own Op in the `stats` namespace! 72d2d7dc14cc0691b67d23aca3089df7aad10bf1

File: scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/QNorm.java (hunk 10 above)
Comment (elevans) [reply scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/QNorm.java]: Agreed! `QNorm` is now an Op in the `stats` namesapce! fefe8bfb2a558ee1bc4a1306963d97e184206e99

File: scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/WtKendallTau.java (hunk 12 above)
Comment (elevans) [reply scijava-ops-image/src/main/java/org/scijava/ops/image/coloc/saca/WtKendallTau.java:48]: I'm against making the `WtKendallTau.java` helper class an Op. It calculates the weighted [`kendall tau rank correlation`](https://en.wikipedia.org/wiki/Kendall_rank_correlation_coefficient) that determines the level/strength of colocalization within a set of data. My understanding here is that this kendall tau class is a specific implementation for SACA and would require a little bit of work to generalize. For now I think this is okay as is.

File: scijava-ops-image/src/main/java/org/scijava/ops/image/simplify/RAIFocusers.java (hunk 13 above)
Comment (elevans) [reply scijava-ops-image/src/main/java/org/scijava/ops/image/simplify/RAIFocusers.java]: Done! 6c9419aa8c9ad58f73e5e40d271c980cae2d1da8

File: scijava-ops-image/src/test/java/org/scijava/ops/image/coloc/saca/SACATest.java (hunk 15 above)
Comment (elevans) [reply scijava-ops-image/src/test/java/org/scijava/ops/image/coloc/saca/SACATest.java]: The test has been split into 2 now! The `coloc.saca.heatmapPValue` Op no longer exists. Its replaced by the `stats.prnom` Op, which has its own test now.

File: scijava-ops-image/src/test/java/org/scijava/ops/image/coloc/saca/SACATest.java (hunk 14 above)
Comment (elevans) [reply scijava-ops-image/src/test/java/org/scijava/ops/image/coloc/saca/SACATest.java:46]: Done!

**Comment by elevans** [comment]
@gselzer Thanks for the awesome review. I think I've replied to everything and addressed all comments. Its ready for another look!

File: docs/ops/doc/examples/saca.rst [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+
+The SACA framework produces three outputs: a *Z*-score heatmap, a p-valye heatmap and a significant pixel value mask. The *Z*-score heatmap
+indicates the relatively colocalization or anti-colocalization strength at a pixel-wise level. The p-value heatmap indicates the p-value
+of the colocalization strength at a pixel-wise level. Finally the significant pixel mask identifies which pixels are significantly colocalized. This
+example script takes advantage of this feature of the SACA framework and utilizes the significant pixel mask as a region of interest to compute
+Pearson's :sup:`4` and Li's :sup:`5` colocalization quotients.
+
+You can download the colocalization dataset `here`_.
+
+.. figure:: https://media.imagej.net/scijava-ops/1.0.0/saca_input.png
```
Comment (gselzer) [inline docs/ops/doc/examples/saca.rst]: With #158 we should rebase this branch over the latest `main`, and update the `media.imagej.net` links to `media.scijava.org`

File: docs/ops/doc/examples/saca.rst [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+        regs = LabelRegions(labeling)
+        coloc_region = regs.getLabelRegion(1)
+        subsample_1 = Regions.sample(coloc_region, channels[0])
+        subsample_2 = Regions.sample(coloc_region, channels[1])
+        pearsons = ops.op("coloc.pearsons").input(subsample_1, subsample_2).apply()
+        li = ops.op("coloc.icq").input(subsample_1, subsample_2).apply()
+        
+        # print Pearson's and Li's results
+        print("pearsons: " + str(pearsons))
+        print("li: " + str(li))
```
Comment (gselzer) [inline docs/ops/doc/examples/saca.rst]: It'd be nice to display the expected values of these returns

File: scijava-ops-image/src/main/java/org/scijava/ops/image/stats/DefaultPNorm.java [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+import org.scijava.ops.spi.Nullable;
+
+/**
+ * @author Edward Evans
+ * @implNote op names='stats.pnorm', priority='100.'
+ */
+
+public final class DefaultPNorm implements
+	Computers.Arity2<RandomAccessibleInterval<DoubleType>, Boolean, RandomAccessibleInterval<DoubleType>>
+{
```
Comment (gselzer) [inline scijava-ops-image/src/main/java/org/scijava/ops/image/stats/DefaultPNorm.java]: Any reason why this has to be a `RAI<DoubleType>`? Could it be a `RAI<O extends RealType<O>>`?

File: scijava-ops-image/src/test/java/org/scijava/ops/image/coloc/saca/SACATest.java [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+		Img<BitType> sigMask = ops.op("create.img").input(zscore_ref, new BitType())
+			.outType(new Nil<Img<BitType>>()
+			{}).apply();
+
+		// run SACA significant pixel mask op
+		ops.op("coloc.saca.sigMask").input(zscore_ref).output(sigMask).compute();
+
+		// get random access and assert results are equal
+		final RandomAccess<BitType> sRA = sigMask.randomAccess();
+		for (int i = 0; i < xPositions.length; i++) {
```
Comment (gselzer) [inline scijava-ops-image/src/test/java/org/scijava/ops/image/coloc/saca/SACATest.java:111]: I often forget about this, but there's a nice method `AbstractOpTest.areCongruent()` to check image equality and also `AbstractOpTest.asArray(Iterable<T> image)` that you could use to avoid manually iterating over the output image.

You don't have to do anything, just an FYI, and also a reminder for me to use it!

File: scijava-ops-image/src/test/java/org/scijava/ops/image/stats/DefaultPNormTest.java [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+	@Test
+	public void testImg() {
+		final int[] xPositions = { 30, 79, 77, 104, 7, 52, 164, 88, 119, 65 };
+		final int[] yPositions = { 30, 36, 80, 79, 139, 102, 77, 41, 142, 118 };
+		final double[] pvalueExpected = { 0.5, 0.9999998132675161, 0.5,
+			0.8861427670894226, 0.5, 0.9999999999828452, 0.5, 0.9960363221624154, 0.5,
+			0.999998128463855 };
+
+		// load Z-score heatmap slice
+		Img<FloatType> zscore = openFloatImg("zscore_test_data.tif");
```
Comment (gselzer) [inline scijava-ops-image/src/test/java/org/scijava/ops/image/stats/DefaultPNormTest.java]: Nothing in the algorithm depends upon the data distribution in the image, correct? If so, then could we:
* Save the 10 data points in a final `double[]`
* Create an `ArrayImg` from that `double[]` using `ArrayImgs.doubles(data, 10)`
* Continue as normal (except you'd have one fewer dimension in your image, which is an added bonus 😉)
That way, we wouldn't have to ship an image with the tests.

**Review by gselzer — CHANGES_REQUESTED** [review]
@elevans thank you so much for addressing my prior review! More Ops = a better framework!

...unfortunately, I'd suggest a couple more things 😅

File: scijava-ops-image/src/test/java/org/scijava/ops/image/coloc/saca/SACATest.java (hunk 19 above)
Comment (elevans) [reply scijava-ops-image/src/test/java/org/scijava/ops/image/coloc/saca/SACATest.java:111]: This would be awesome to use, but unfortunately for me I'm extending `AbstractColocalisationTest` and not `AbstractOpTest` where `areCongruent()` lives. I'm just going to stick with this manual way for now :D.

File: docs/ops/doc/examples/saca.rst (hunk 16 above)
Comment (elevans) [reply docs/ops/doc/examples/saca.rst]: Done!

File: docs/ops/doc/examples/saca.rst (hunk 17 above)
Comment (elevans) [reply docs/ops/doc/examples/saca.rst]: Done! I agree, this was an over sight.

File: scijava-ops-image/src/main/java/org/scijava/ops/image/stats/DefaultPNorm.java (hunk 18 above)
Comment (elevans) [reply scijava-ops-image/src/main/java/org/scijava/ops/image/stats/DefaultPNorm.java]: There is not reason and in fact with your help we made this Op better be allowing it work on pixels instead of just RAIs!

File: scijava-ops-image/src/test/java/org/scijava/ops/image/stats/DefaultPNormTest.java (hunk 20 above)
Comment (elevans) [reply scijava-ops-image/src/test/java/org/scijava/ops/image/stats/DefaultPNormTest.java]: Done! With this change I was able to remove the need to read an image and also ship an image in the tests!

**Comment by elevans** [comment]
@gselzer This PR is ready for its second round of review. Thanks for all the help!

**Review by gselzer — APPROVED** [review]
